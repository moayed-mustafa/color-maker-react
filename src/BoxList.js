import React, { useState } from 'react'
import Box from './Box'
import { v4 as uuid } from 'uuid'
import BoxForm from './BoxForm'


const BoxList = (props) => {

    // set the state
    const INITIAL_STATE = [
        {id: uuid(), color: 'red', width: 200, height:200},
        {id: uuid(), color: 'navy', width: 200, height:200},
        {id: uuid(), color: 'green', width: 200, height:200},

    ]
    const [boxes, setBoxes] = useState(INITIAL_STATE)
    // delete a box
    const handleDelete = (id) => {
        let newBoxes = boxes.filter(box => box.id !== id)
        setBoxes(newBoxes)
    }
    // add a box
    const handleAdding = (boxes) => {
        setBoxes((data) =>[...data, {id:uuid(), ...boxes}] )
    }


    return (
        // this should take color, width , height
            <>
            <BoxForm submit={handleAdding} />
            <div>
                {boxes.map((box) =>
                    <Box
                        key={box.id}
                        color={box.color}
                        width={box.width}
                        height={box.height}
                        handler={() => handleDelete(box.id)}
                        id={box.id}
                    />)}
            </div>

            </>


    )
}

export default BoxList