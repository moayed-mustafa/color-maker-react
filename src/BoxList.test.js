import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'


import BoxList from './BoxList.js'

// smoke test
it('should render without crashing', () => {
    render(<BoxList/>)
})

// snapshot test
it('should match an existing snapshot', () => {
    const { asFragment } = render(<BoxList />)

    expect(asFragment).toMatchSnapshot()
})

it('should be able to add a box', () => {
    const { queryByText, queryByRole, getByPlaceholderText } = render(<BoxList />)
    const button = queryByText('Add')
    expect(queryByRole('pink-box')).not.toBeInTheDocument()
    const inputs = [
         'width',
         'height',
         'color'
    ]

    const values = [200, 200, 'pink']
    for (let i = 0; i < inputs.length; i++){
        fireEvent.change(getByPlaceholderText(inputs[i]), {target:{value: values[i]}})
    }
    fireEvent.click(button)
    expect(queryByRole('pink-box')).toBeInTheDocument()


})

it('should be able to delete a box', () => {
    // todo: this is a bit quirky since I know that the red box is the first one, what if I did not?

    const { queryAllByText,queryByRole } = render(<BoxList />)
    const button = queryAllByText('X')
    expect(queryByRole('red-box')).toBeInTheDocument()
    fireEvent.click(button[0])
    expect(queryByRole('red-box')).not.toBeInTheDocument()
})