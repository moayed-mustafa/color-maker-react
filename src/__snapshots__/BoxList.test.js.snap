// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match an existing snapshot 1`] = `[Function]`;
